File: libopae/plugins/xfpga/properties.c
// Copyright(c) 2017-2018, Intel Corporation
//
// Redistribution  and  use  in source  and  binary  forms,  with  or  without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of  source code  must retain the  above copyright notice,
//   this list of conditions and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright notice,
//   this list of conditions and the following disclaimer in the documentation
//   and/or other materials provided with the distribution.
// * Neither the name  of Intel Corporation  nor the names of its contributors
//   may be used to  endorse or promote  products derived  from this  software
//   without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif // HAVE_CONFIG_H

#include <opae/properties.h>

#include "safe_string/safe_string.h"

#include "xfpga.h"
#include "common_int.h"
#include "props.h"
#include "error_int.h"


fpga_result __FPGA_API__
xfpga_fpgaGetPropertiesFromHandle(fpga_handle handle, fpga_properties *prop)
{
	struct _fpga_handle *_handle = (struct _fpga_handle *)handle;
	fpga_result result = FPGA_OK;
	int err = 0;

	result = handle_check_and_lock(_handle);
	if (result)
		return result;

	result = xfpga_fpgaGetProperties(_handle->token, prop);

	err = pthread_mutex_unlock(&_handle->lock);
	if (err) {
		FPGA_ERR("pthread_mutex_unlock() failed: %S", strerror(err));
	}

	return result;
}

fpga_result __FPGA_API__ xfpga_fpgaGetProperties(fpga_token token,
						 fpga_properties *prop)
{
	struct _fpga_properties *_prop = NULL;
	fpga_result result = FPGA_OK;

	ASSERT_NOT_NULL(prop);

	result = fpgaGetProperties(NULL, (fpga_properties *)&_prop);

	ASSERT_RESULT(result);

	if (token) {
		result = xfpga_fpgaUpdateProperties(token, _prop);
		if (result != FPGA_OK)
			goto out_free;
	}

	*prop = (fpga_properties)_prop;
	return result;

out_free:
	free(_prop);
	return result;
}

fpga_result __FPGA_API__ xfpga_fpgaUpdateProperties(fpga_token token,
						    fpga_properties prop)
{
	struct _fpga_token *_token = (struct _fpga_token *)token;
	struct _fpga_properties *_prop = (struct _fpga_properties *)prop;

	struct _fpga_properties _iprop;

	char spath[SYSFS_PATH_MAX];
	char *p;
	int s, b, d, f;
	int res;
	errno_t e;
	int err = 0;
	int resval = 0;
	uint64_t value = 0;

	pthread_mutex_t lock;

	fpga_result result = FPGA_INVALID_PARAM;

	ASSERT_NOT_NULL(token);
	if (_token->magic != FPGA_TOKEN_MAGIC) {
		FPGA_MSG("Invalid token");
		return FPGA_INVALID_PARAM;
	}

	ASSERT_NOT_NULL(_prop);
	if (_prop->magic != FPGA_PROPERTY_MAGIC) {
		FPGA_MSG("Invalid properties object");
		return FPGA_INVALID_PARAM;
	}

	// clear fpga_properties buffer
	memset_s(&_iprop, sizeof(struct _fpga_properties), 0);
	_iprop.magic = FPGA_PROPERTY_MAGIC;

	// The input token is either for an FME or an AFU.
	// Go one level back to get to the dev.

	e = strncpy_s(spath, sizeof(spath), _token->sysfspath,
		      sizeof(_token->sysfspath));
	if (EOK != e) {
		FPGA_ERR("strncpy_s failed");
		return FPGA_EXCEPTION;
	}

	p = strrchr(spath, '/');
	ASSERT_NOT_NULL_MSG(p, "Invalid token sysfs path");

	*p = 0;

	p = strstr(_token->sysfspath, FPGA_SYSFS_AFU);
	if (NULL != p) {
		// AFU
		result = sysfs_get_guid(_token, FPGA_SYSFS_AFU_GUID,
			 _iprop.guid);
		if (FPGA_OK != result)
			return result;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_GUID);

		_iprop.parent = (fpga_token)token_get_parent(_token);
		if (NULL != _iprop.parent)
			SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_PARENT);

		_iprop.objtype = FPGA_ACCELERATOR;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_OBJTYPE);

		res = open(_token->devpath, O_RDWR);
		if (-1 == res) {
			_iprop.u.accelerator.state = FPGA_ACCELERATOR_ASSIGNED;
		} else {
			close(res);
			_iprop.u.accelerator.state =
				FPGA_ACCELERATOR_UNASSIGNED;
		}
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_ACCELERATOR_STATE);

		_iprop.u.accelerator.num_mmio = 2;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_NUM_MMIO);

		_iprop.u.accelerator.num_interrupts = 0;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_NUM_INTERRUPTS);
	}

	p = strstr(_token->sysfspath, FPGA_SYSFS_FME);
	if (NULL != p) {
		// FME
		_iprop.objtype = FPGA_DEVICE;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_OBJTYPE);
		// get bitstream id
		result = sysfs_get_interface_id(_token, _iprop.guid);	
		if (FPGA_OK != result)
			return result;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_GUID);

		resval = sysfs_parse_attribute64(_token->sysfspath,
			FPGA_SYSFS_NUM_SLOTS, &value);
		if (resval != 0) {
			return FPGA_NOT_FOUND;
		}
		_iprop.u.fpga.num_slots = (uint32_t)value;
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_NUM_SLOTS);

		resval = sysfs_parse_attribute64(_token->sysfspath,
			FPGA_SYSFS_BITSTREAM_ID, &_iprop.u.fpga.bbs_id);
		if (resval != 0) {
			return FPGA_NOT_FOUND;
		}
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_BBSID);

		_iprop.u.fpga.bbs_version.major =
			FPGA_BBS_VER_MAJOR(_iprop.u.fpga.bbs_id);
		_iprop.u.fpga.bbs_version.minor =
			FPGA_BBS_VER_MINOR(_iprop.u.fpga.bbs_id);
		_iprop.u.fpga.bbs_version.patch =
			FPGA_BBS_VER_PATCH(_iprop.u.fpga.bbs_id);
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_BBSVERSION);
	}

	result = sysfs_sbdf_from_path(spath, &s, &b, &d, &f);
	if (result)
		return result;

	_iprop.segment = (uint16_t)s;
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_SEGMENT);

	_iprop.bus = (uint8_t)b;
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_BUS);

	_iprop.device = (uint8_t)d;
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_DEVICE);

	_iprop.function = (uint8_t)f;
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_FUNCTION);

	// only set socket id if we have it on sysfs
	result = sysfs_get_socket_id(_token->device_instance,
		_token->subdev_instance,
		&_iprop.socket_id);
	if (0 == result)
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_SOCKETID);

	result = sysfs_objectid_from_path(_token->sysfspath, &_iprop.object_id);
	if (0 == result)
		SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_OBJECTID);

	// read the vendor and device ID from the 'device' path
	uint32_t x = 0;
	char vendorpath[SYSFS_PATH_MAX];
	snprintf_s_s(vendorpath, SYSFS_PATH_MAX, "%s/../device/vendor",
		     _token->sysfspath);
	result = sysfs_read_u32(vendorpath, &x);
	if (result != FPGA_OK)
		return result;
	_iprop.vendor_id = (uint16_t)x;
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_VENDORID);

	char devicepath[SYSFS_PATH_MAX];
	snprintf_s_s(devicepath, SYSFS_PATH_MAX, "%s/../device/device",
		     _token->sysfspath);
	result = sysfs_read_u32(devicepath, &x);
	if (result != FPGA_OK)
		return result;
	_iprop.device_id = (uint16_t)x;
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_DEVICEID);

	char errpath[SYSFS_PATH_MAX];
	snprintf_s_s(errpath, SYSFS_PATH_MAX, "%s/errors", _token->sysfspath);
	_iprop.num_errors = count_error_files(errpath);
	SET_FIELD_VALID(&_iprop, FPGA_PROPERTY_NUM_ERRORS);

	if (pthread_mutex_lock(&_prop->lock)) {
		FPGA_MSG("Failed to lock properties mutex");
		return FPGA_EXCEPTION;
	}

	lock = _prop->lock;
	*_prop = _iprop;
	_prop->lock = lock;

	err = pthread_mutex_unlock(&_prop->lock);
	if (err)
		FPGA_ERR("pthread_mutex_unlock() failed: %s", strerror(err));

	return FPGA_OK;
}

File: python/opae.http/opae/http/properties.py
#!/usr/bin/env python3
# Copyright(c) 2023, Intel Corporation
#
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of  source code  must retain the  above copyright notice,
#   this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
#   may be used to  endorse or promote  products derived  from this  software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import json

import opae.http.constants as constants
from opae.http.version import version
from opae.http.conversion import to_json_obj


class tracks_valid_fields():
    def __init__(self):
        self.__dict__['valid_fields'] = 0

    def field_is_valid(self, f):
        valid_fields = self.__dict__['valid_fields']
        return ((valid_fields >> f) & 1) != 0

    def set_field_valid(self, f):
        self.__dict__['valid_fields'] |= (1 << f)


class fpga_properties(tracks_valid_fields):
    NUM_SLOTS_IDX = 32
    BBSID_IDX = 33
    BBSVERSION_IDX = 34

    attrs = {'num_slots': NUM_SLOTS_IDX,
             'bbs_id': BBSID_IDX,
             'bbs_version': BBSVERSION_IDX,
            }

    def __init__(self, parent):
        super().__init__()
        self.__dict__['_parent'] = parent

    def set_field_valid(self, f):
        super().set_field_valid(f)
        parent = self.__dict__['_parent']
        parent.set_field_valid(f)

    def __getattr__(self, attr):
        if attr in fpga_properties.attrs:
            if self.field_is_valid(fpga_properties.attrs[attr]):
                return self.__dict__[attr]
            else:
                raise AttributeError(attr + ' is not set on fpga_properties')

    def __setattr__(self, attr, value):
        if attr in fpga_properties.attrs:
            self.__dict__[attr] = value
            self.set_field_valid(fpga_properties.attrs[attr])
        else:
            raise AttributeError(attr + ' is not a valid fpga_properties attr')

    def to_json_obj(self):
        d = {}
        transforms = {'bbs_id': str,
                      'bbs_version': to_json_obj}
        for attr, idx in fpga_properties.attrs.items():
            if self.field_is_valid(idx):
                if attr in transforms:
                    d[attr] = transforms[attr](self.__dict__[attr])
                else:
                    d[attr] = self.__dict__[attr]
        return d

    @staticmethod
    def resolve_aliases(d):
        aliases = {'num_slots':   ['numSlots'],
                   'bbs_id':      ['bbsId'],
                   'bbs_version': ['bbsVersion'],
                  }
        c = d.copy()
        for attr, alias_list in aliases.items():
            if attr in c:
                continue
            for a in alias_list:
                if a in c:
                    c[attr] = c[a]
                    del c[a]
        return c

    @staticmethod
    def from_json_obj(jobj, parent):
        props = fpga_properties(parent)
        transforms = {'bbs_version': version.from_json_obj}
        for k, v in fpga_properties.resolve_aliases(jobj).items():
            if k in transforms:
                setattr(props, k, transforms[k](v))
            else:
                setattr(props, k, v)
        return props


class accelerator_properties(tracks_valid_fields):
    ACCELERATOR_STATE_IDX = 32
    NUM_MMIO_IDX = 33
    NUM_INTERRUPTS_IDX = 34

    attrs = {'state': ACCELERATOR_STATE_IDX,
             'num_mmio': NUM_MMIO_IDX,
             'num_interrupts': NUM_INTERRUPTS_IDX,
            }

    def __init__(self, parent):
        super().__init__()
        self.__dict__['_parent'] = parent

    def set_field_valid(self, f):
        super().set_field_valid(f)
        parent = self.__dict__['_parent']
        parent.set_field_valid(f)

    def __getattr__(self, attr):
        if attr in accelerator_properties.attrs:
            if self.field_is_valid(accelerator_properties.attrs[attr]):
                return self.__dict__[attr]
            else:
                raise AttributeError(attr + ' is not set on accelerator_properties')

    def __setattr__(self, attr, value):
        if attr in accelerator_properties.attrs:
            self.__dict__[attr] = value
            self.set_field_valid(accelerator_properties.attrs[attr])
        else:
            raise AttributeError(attr + ' is not a valid accelerator_properties attr')

    def to_json_obj(self):
        d = {}
        transforms = {'state': constants.fpga_accelerator_state_to_str}
        for attr, idx in accelerator_properties.attrs.items():
            if self.field_is_valid(idx):
                if attr in transforms:
                    d[attr] = transforms[attr](self.__dict__[attr])
                else:
                    d[attr] = self.__dict__[attr]
        return d

    @staticmethod
    def resolve_aliases(d):
        aliases = {'num_mmio':       ['numMmio'],
                   'num_interrupts': ['numInterrupts'],
                  }
        c = d.copy()
        for attr, alias_list in aliases.items():
            if attr in c:
                continue
            for a in alias_list:
                if a in c:
                    c[attr] = c[a]
                    del c[a]
        return c

    @staticmethod
    def from_json_obj(jobj, parent):
        props = accelerator_properties(parent)
        transforms = {'state': constants.fpga_accelerator_state_from_str}
        for k, v in accelerator_properties.resolve_aliases(jobj).items():
            if k in transforms:
                setattr(props, k, transforms[k](v))
            else:
                setattr(props, k, v)
        return props


class properties(tracks_valid_fields):
    MAGIC = 0x4650474150524f50

    PARENT_IDX = 0
    OBJTYPE_IDX = 1
    SEGMENT_IDX = 2
    BUS_IDX = 3
    DEVICE_IDX = 4
    FUNCTION_IDX = 5
    SOCKETID_IDX = 6
    VENDORID_IDX = 7
    DEVICEID_IDX = 8
    GUID_IDX = 9
    OBJECTID_IDX = 10
    NUM_ERRORS_IDX = 11
    INTERFACE_IDX = 12
    SUB_VENDORID_IDX = 13
    SUB_DEVICEID_IDX = 14
    HOSTNAME_IDX = 15

    attrs = {'guid': GUID_IDX,
             'parent': PARENT_IDX,
             'objtype': OBJTYPE_IDX,
             'segment': SEGMENT_IDX,
             'bus': BUS_IDX,
             'device': DEVICE_IDX,
             'function': FUNCTION_IDX,
             'socket_id': SOCKETID_IDX,
             'object_id': OBJECTID_IDX,
             'vendor_id': VENDORID_IDX,
             'device_id': DEVICEID_IDX,
             'num_errors': NUM_ERRORS_IDX,
             'interface': INTERFACE_IDX,
             'subsystem_vendor_id': SUB_VENDORID_IDX,
             'subsystem_device_id': SUB_DEVICEID_IDX,
             'hostname': HOSTNAME_IDX,
            }

    def __init__(self):
        super().__init__()
        self.__dict__['fpga'] = fpga_properties(self)
        self.__dict__['accelerator'] = accelerator_properties(self)

    def __getattr__(self, attr):
        objtype = self.__dict__.get('objtype', constants.FPGA_ACCELERATOR + 1)
        if attr in properties.attrs:
            if self.field_is_valid(properties.attrs[attr]):
                return self.__dict__[attr]
            else:
                raise AttributeError(attr + ' is not set')
        elif ((attr == 'fpga') and
              (objtype == constants.FPGA_DEVICE)):
                return self.__dict__['fpga']
        elif ((attr == 'accelerator') and
              (objtype == constants.FPGA_ACCELERATOR)):
                return self.__dict__['accelerator']
        raise AttributeError(attr + ' is not a valid property')

    def __setattr__(self, attr, value):
        objtype = self.__dict__.get('objtype', constants.FPGA_ACCELERATOR + 1)
        valid = False
        index = -1
        if attr in properties.attrs:
            valid = True
            index = properties.attrs[attr]
        elif ((attr == 'fpga') and
              (objtype == constants.FPGA_DEVICE)):
            return setattr(self.__dict__['fpga'], value)
        elif ((attr == 'accelerator') and
              (objtype == constants.FPGA_ACCELERATOR)):
            return setattr(self.__dict__['accelerator'], value)

        if valid:
            self.__dict__[attr] = value
            self.set_field_valid(index)
        else:
            raise AttributeError(attr + ' is not a valid property')

    def to_json_obj(self):
        d = {}
        d['magic'] = str(properties.MAGIC)
        d['valid_fields'] = str(self.__dict__['valid_fields'])
        transforms = {'objtype': constants.fpga_objtype_to_str,
                      'interface': constants.fpga_interface_to_str,
                      'object_id': str,
                     }
        objtype = self.__dict__.get('objtype', constants.FPGA_ACCELERATOR + 1)
        for attr, idx in properties.attrs.items():
            if self.field_is_valid(idx):
                if attr in transforms:
                    d[attr] = transforms[attr](self.__dict__[attr])
                else:
                    d[attr] = self.__dict__[attr]
        if objtype == constants.FPGA_DEVICE:
            fpga = self.__dict__['fpga'].to_json_obj()
            if len(fpga):
                d['fpga'] = fpga
        elif objtype == constants.FPGA_ACCELERATOR:
            accelerator = self.__dict__['accelerator'].to_json_obj()
            if len(accelerator):
                d['accelerator'] = accelerator
        return d

    def to_json_str(self):
        return json.dumps(self.to_json_obj())

    @staticmethod
    def resolve_aliases(d):
        aliases = {'guid':                [],
                   'parent':              [],
                   'objtype':             [],
                   'segment':             [],
                   'bus':                 [],
                   'device':              [],
                   'function':            [],
                   'socket_id':           ['socketId'],
                   'object_id':           ['objectId'],
                   'vendor_id':           ['vendorId'],
                   'device_id':           ['deviceId'],
                   'num_errors':          ['numErrors'],
                   'interface':           [],
                   'subsystem_vendor_id': ['subsystemVendorId'],
                   'subsystem_device_id': ['subsystemDeviceId'],
                   'hostname':            [],
                  }
        c = d.copy()
        for attr, alias_list in aliases.items():
            if attr in c:
                continue
            for a in alias_list:
                if a in c:
                    c[attr] = c[a]
                    del c[a]
        return c

    @staticmethod
    def from_json_obj(jobj):
        props = properties()
        props.__dict__['magic'] = properties.MAGIC

        valid_fields = jobj.get('valid_fields')
        if valid_fields is None:
            valid_fields = jobj.get('validFields')
        props.__dict__['valid_fields'] = int(valid_fields)

        objtype = jobj.get('objtype', constants.FPGA_ACCELERATOR + 1)
        if objtype in [constants.FPGA_DEVICE, constants.FPGA_ACCELERATOR]:
            setattr(props, 'objtype', constants.fpga_objtype_from_str(objtype))

        transforms = {'objtype': constants.fpga_objtype_from_str,
                      'interface': constants.fpga_interface_from_str,
                      'object_id': int,
                     }
        skip = ['magic', 'valid_fields', 'validFields']
        for k, v in properties.resolve_aliases(jobj).items():
            if k in skip:
                continue
            if k == 'fpga':
                props.__dict__['fpga'] = fpga_properties.from_json_obj(v, props)
            elif k == 'accelerator':
                props.__dict__['accelerator'] = accelerator_properties.from_json_obj(v, props)
            else:
                if k in transforms:
                    setattr(props, k, transforms[k](v))
                else:
                    setattr(props, k, v)
        return props
